L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc

```diff
@@ -148,65 +172,172 @@ void TTStubBuilder< Ref_Phase2TrackerDigi_ >::produce( edm::Event& iEvent, const
+		tempTTStub2.setHardwareBend( tempTTStub.getHardwareBend() );
+
+		tempInner.push_back( *(tempTTStub2.getClusterRef(0)) );
+		tempOuter.push_back( *(tempTTStub2.getClusterRef(1)) );
+		tempAccepted.push_back( tempTTStub2 );
+		FEreject = true;
+	      }
+	    }
+	  }
+
+	  // End of the MPA/CBC loop
+
+	  // If the stub has been already thrown out, there is no reason to include it into the CIC stream
+	  if (FEreject) continue;
+	  
+	  (isPS)
+	    ? maxStubs = maxStubs_PS_CIC_5
+	    : maxStubs = maxStubs_2S_CIC_5;
+
+	  if (is10G_PS) maxStubs = maxStubs_PS_CIC_10;
+
+	  if ( moduleStubs_CIC.find( CIC_chip ) == moduleStubs_CIC.end() ) /// Already a stub for this ASIC?
           {
             /// No, so new entry
             std::vector< TTStub< Ref_Phase2TrackerDigi_ > > tempStubs(1,tempTTStub);
-            //tempStubs.clear();
-            //tempStubs.push_back( tempTTStub );
-            moduleStubs.insert( std::pair< int, std::vector< TTStub< Ref_Phase2TrackerDigi_ > > >( chip, tempStubs ) );
+            moduleStubs_CIC.insert( std::pair< int, std::vector< TTStub< Ref_Phase2TrackerDigi_ > > >( CIC_chip, tempStubs ) );
+
+	    tempInner.push_back( *(tempTTStub.getClusterRef(0)) );
+            tempOuter.push_back( *(tempTTStub.getClusterRef(1)) );
+            tempAccepted.push_back( tempTTStub ); // The stub is added
           }
           else
           {
-            /// Already existing entry
-            moduleStubs[chip].push_back( tempTTStub );
+	    bool CIC_reject=true;
+
+	    if ( moduleStubs_CIC[CIC_chip].size() < maxStubs )
+	    {
+	      moduleStubs_CIC[CIC_chip].push_back( tempTTStub ); //We add the new stub
+	      tempInner.push_back( *(tempTTStub.getClusterRef(0)) );
+	      tempOuter.push_back( *(tempTTStub.getClusterRef(1)) );
+	      tempAccepted.push_back( tempTTStub ); // The stub is added
+	    }
+	    else
+	    {
+	      moduleStubs_CIC[CIC_chip].push_back( tempTTStub ); //We add the new stub
+
+	      /// Sort them by |bend| and pick up only the first N.
+	      bendMap.clear();
+	      bendMap.reserve(moduleStubs_CIC[CIC_chip].size());
+	    
+	      for ( unsigned int i = 0; i < moduleStubs_CIC[CIC_chip].size(); ++i )
+	      {
+		bendMap.push_back( std::pair< unsigned int, double >( i, moduleStubs_CIC[CIC_chip].at(i).getTriggerBend() ) );
+	      }
+
+	      std::sort( bendMap.begin(), bendMap.end(), TTStubBuilder< Ref_Phase2TrackerDigi_ >::SortStubBendPairs );
+	      
+	      // bendMap contains link over all the stubs included in moduleStubs_CIC[CIC_chip]
+	      
+	      for ( unsigned int i = 0; i < maxStubs; ++i ) 
+	      {
+		// The stub we have added is among the first ones, add it
+		if (bendMap[i].first==moduleStubs_CIC[CIC_chip].size()-1)
+		{
+		  CIC_reject=false;
+		}
+	      }
+
+	      if (CIC_reject) // The stub added does not pass the cut
+	      {
+		TTStub< Ref_Phase2TrackerDigi_ > tempTTStub2( tempTTStub.getDetId() );
```

Comment:
try to reduce code duplication here as well (very similar to above cases)

Reply:
Not done

Reply:
Why not? The code appears identical except for 500 vs 1000

Reply:
I have added some extra-simplification, but will not create a new function simply for that.